src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(stateCopy, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const i of action.keysToRemove) {
+          if (stateCopy[i]) {
+            delete stateCopy[i];
+          }
+        }
+        break;
+
+      case 'clear':
+        stateCopy = {};
+        break;
+
+      default:
+    }
+    stateHistory.push({ ...stateCopy });
```

Comment:
You are pushing the state copy to the history after each action, which is correct. However, to follow the 'Don't Repeat Yourself' principle, consider moving this operation outside of the switch statement, as it is repeated in every case.